File: src/com/l2kt/gameserver/network/clientpackets/RequestGetOnVehicle.java
package com.l2kt.gameserver.network.clientpackets;

import com.l2kt.gameserver.data.manager.BoatManager;
import com.l2kt.gameserver.model.actor.Boat;
import com.l2kt.gameserver.model.actor.instance.Player;
import com.l2kt.gameserver.network.serverpackets.ActionFailed;
import com.l2kt.gameserver.network.serverpackets.GetOnVehicle;

public final class RequestGetOnVehicle extends L2GameClientPacket
{
	private int _boatId;
	private int _x;
	private int _y;
	private int _z;
	
	@Override
	protected void readImpl()
	{
		_boatId = readD();
		_x = readD();
		_y = readD();
		_z = readD();
	}
	
	@Override
	protected void runImpl()
	{
		final Player activeChar = getClient().getActiveChar();
		if (activeChar == null)
			return;
		
		Boat boat;
		if (activeChar.isInBoat())
		{
			boat = activeChar.getBoat();
			if (boat.getObjectId() != _boatId)
			{
				sendPacket(ActionFailed.STATIC_PACKET);
				return;
			}
		}
		else
		{
			boat = BoatManager.getInstance().getBoat(_boatId);
			if (boat == null || boat.isMoving() || !activeChar.isInsideRadius(boat, 1000, true, false))
			{
				sendPacket(ActionFailed.STATIC_PACKET);
				return;
			}
		}
		
		activeChar.getBoatPosition().set(_x, _y, _z, activeChar.getHeading());
		activeChar.setBoat(boat);
		activeChar.broadcastPacket(new GetOnVehicle(activeChar.getObjectId(), boat.getObjectId(), _x, _y, _z));
		
		activeChar.setXYZ(boat.getX(), boat.getY(), boat.getZ());
		activeChar.revalidateZone(true);
	}
}

File: src/com/l2kt/gameserver/network/clientpackets/RequestGetOnVehicle.kt
package com.l2kt.gameserver.network.clientpackets

import com.l2kt.gameserver.data.manager.BoatManager
import com.l2kt.gameserver.model.actor.Boat
import com.l2kt.gameserver.network.serverpackets.ActionFailed
import com.l2kt.gameserver.network.serverpackets.GetOnVehicle

class RequestGetOnVehicle : L2GameClientPacket() {
    private var _boatId: Int = 0
    private var _x: Int = 0
    private var _y: Int = 0
    private var _z: Int = 0

    override fun readImpl() {
        _boatId = readD()
        _x = readD()
        _y = readD()
        _z = readD()
    }

    override fun runImpl() {
        val activeChar = client.activeChar ?: return

        val boat: Boat?
        if (activeChar.isInBoat) {
            boat = activeChar.boat
            if (boat!!.objectId != _boatId) {
                sendPacket(ActionFailed.STATIC_PACKET)
                return
            }
        } else {
            boat = BoatManager.getBoat(_boatId)
            if (boat == null || boat.isMoving || !activeChar.isInsideRadius(boat, 1000, true, false)) {
                sendPacket(ActionFailed.STATIC_PACKET)
                return
            }
        }

        activeChar.boatPosition.set(_x, _y, _z, activeChar.heading)
        activeChar.boat = boat
        activeChar.broadcastPacket(GetOnVehicle(activeChar.objectId, boat.objectId, _x, _y, _z))

        activeChar.setXYZ(boat.x, boat.y, boat.z)
        activeChar.revalidateZone(true)
    }
}